spmd/tensor/parallel/style.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import torch
+from dataclasses import dataclass
+from typing import Callable, Union, Optional
+from spmd.tensor import (
+    DTensor,
+    Shard,
+    Replicate,
+    DeviceMesh,
+)
+
+
+@dataclass
+class ParallelStyle(object):
+    """
+    The parallel style user wants the module or submodule to be sharded.
+    Users can extend this class to build their own parallel style with customized input/output preparations.
+    """
+
+    _prepare_input: Optional[
+        Callable[
+            [Union[torch.Tensor, DTensor], Optional[DeviceMesh], Optional[int]],
```

Comment:
Upon checking, this is fine.

Optional[X] is equivalent to X | None (or Union[X, None]). So let's keep Optional[int].

https://docs.python.org/3/library/typing.html